src/base/static/state/ducks/map.js

```diff
@@ -1,311 +1,581 @@
+          // though, so it would be great if react-map-gl could just
+          // incorporate the scroll zoom around center option natively.
+          // See: https://github.com/uber/react-map-gl/issues/515
+          latitude:
+            action.payload.scrollZoomAroundCenter &&
+            action.payload.viewport.zoom !== state.viewport.zoom
+              ? state.viewport.latitude
+              : action.payload.viewport.latitude
+                ? action.payload.viewport.latitude
+                : state.viewport.latitude,
+          longitude:
+            action.payload.scrollZoomAroundCenter &&
+            action.payload.viewport.zoom !== state.viewport.zoom
+              ? state.viewport.longitude
+              : action.payload.viewport.longitude
+                ? action.payload.viewport.longitude
+                : state.viewport.longitude,
         },
       };
-    case UPDATE_LAYERS_HIDDEN:
-      newStatuses = action.payload.reduce((memo, layerId) => {
-        return {
-          ...memo,
-          [layerId]: {
-            ...state.layerStatuses[layerId],
-            isVisible: false,
-          },
-        };
-      }, {});
-
+    case LOAD_STYLE_AND_METADATA:
       return {
         ...state,
-        layerStatuses: {
-          ...state.layerStatuses,
-          ...newStatuses,
+        style: {
+          ...state.style,
+          ...action.payload.style,
+        },
+        layerGroupsMetadata: {
+          ...state.layerGroupsMetadata,
+          ...action.payload.layerGroupsMetadata,
         },
+        sourcesMetadata: {
+          ...state.sourcesMetadata,
+          ...action.payload.sourcesMetadata,
+        },
+        basemapLayerIds: action.payload.basemapLayerIds,
+        interactiveLayerIds: action.payload.interactiveLayerIds,
       };
-    case SET_LAYER_LOAD_STATUS:
+    case UPDATE_SOURCE_LOAD_STATUS:
       return {
         ...state,
-        layerStatuses: {
-          ...state.layerStatuses,
-          [action.payload.id]: {
-            ...state.layerStatuses[action.payload.id],
+        sourcesMetadata: {
+          ...state.sourcesMetadata,
+          [action.payload.sourceId]: {
+            ...state.sourcesMetadata[action.payload.sourceId],
             loadStatus: action.payload.loadStatus,
           },
         },
       };
-    case ACTIVATE_FEATURE_FILTER:
+    case UPDATE_LAYER_GROUP_LOAD_STATUS:
       return {
         ...state,
-        updatingFilterGroupId: action.payload.groupId,
-        updatingFilterTargetLayer: action.payload.targetLayer,
-        featureFilters: state.featureFilters.concat(action.payload),
+        layerGroupsMetadata: {
+          ...state.layerGroupsMetadata,
+          [action.payload.groupId]: {
+            ...state.layerGrouspsMetadata[action.payload.groupId],
+            loadStatus: action.payload.loadStatus,
+          },
+        },
       };
-    case DEACTIVATE_FEATURE_FILTER:
+    case UPDATE_LAYER_GROUP_VISIBILITY:
```

Comment:
This visibility update is pretty complex but still an improvement over what we had before, I think.

Reply:
That's an interesting idea. The main issue here is that the `style` object is literally a mapbox style object, stored in Redux. I don't think we can add extra stuff to data structures within it (like `layers` and `sources`) because Mapbox will complain that we have an ill-formatted style object. Thus the need for separate "metadata" tracking data structures.

But I agree it seems overly complex. I'll think some more about an alternate architecture.

Reply:
> The main issue here is that the `style` object is literally a mapbox style object, stored in Redux.

I left a comment above that references this thread, which explains why I don't think we need to have our Redux store mirror the mapbox `style` object.